doc/designs/lms.md

```diff
@@ -0,0 +1,243 @@
+/* Alias: since the function signature is identical */
+#define EVP_DigestVerifyHBSUpdate(a,b,c) EVP_DigestVerifyUpdate(a,b,c)
+
+__owur int EVP_DigestVerifyHBSFinal(EVP_MD_CTX *ctx);
+```
+
+Proposal 2:
+
+Add a Init_ex function and leave the Final the same.
+
+i.e.
+EVP_DigestVerifyInit_ex3(..., const unsigned char *sig, size_t siglen);
+
+For the EVP_DigestVerifyFinal pass a NULL signature.
+The signature length must be zero in this case or an error will occur.
+If a signature is passed should it just ignore the value?
+
+Proposal 3:
+
+Pass the signature after the init via an OSSL_PARAM.
+
+
+Hash Function Selection
+-----------------------
+
+- During the DigestVerifyInit()the passed in digest name is not required
+since the LMS/HSS public key and HSS signature contain the digest to use.
+If specified, the digest name must match the one used by the HSS public key and
+signature, otherwise an error will occur.
+- Section 4 of [2] states that the same hash function MUST be used everywhere
+for OTS and LMS at all HSS levels (this means the terms 'm' and 'n' are the
+same)). ([1] only states that the hashes SHOULD be the same).
+This requires error checking when the HSS signature data is decoded.
+- Hash Functions for SHAKE and truncated SHA256 are not specified in [1],
+but are part of [2] so they should be supported.
+For truncated SHAKE256/192 support this will require the OSSL_DIGEST_PARAM_XOFLEN
+parameter to be set.
+
+Threading
+---------
+
+For HSS, multi-level trees are used for verification,
+and there is a verify process involving many hashes for each of the (1..8)
+levels, with only the final level using the input message (and the message is
+possibly streamed).
+All the other levels are one shot verify operations that must all return true in
+order to pass verification. These one-shot operations may be done in parallel
+using the new internal thread API calls (introduced for ARGON).
+When streaming only the final part of the validation involves the message,
+and this is done using a init, update and final in the main thread.
+After this the other threads must finish before verification can end.
+If threads are not supported, then a single threaded call is used instead.
+By default the thread count will be one.
+A parameter to set the threads can be done via a OSSL_PARAM.
+For one shot verification threading can still be used for the top levels.
+
+Provider Support
+----------------
+
+dupctx should not be supported by the provider since this would not apply to LMS.
+
+In order to support the new API calls there should be 2 additional dispatch
+functions..
+
+OSSL_CORE_MAKE_FUNC(int, signature_digest_verify_pbs_init,
+                    (void *ctx, const char *mdname, void *provkey,
+                     const OSSL_PARAM params[],
+                     const unsigned char *sig, size_t siglen))
+OSSL_CORE_MAKE_FUNC(int, signature_digest_verify_pbs_final,
+                    (void *ctx))
```

Comment:
Note: It should be possible to implement this _init_with_signature_ function in terms of the existing APIs for those algorithms that do not have this new provided function. We would just cache the signature and use it in the final call. That's another reason why the EVP function should not be named with HBS in the name as this can be a fully generic call supported even with older signature algorithms on older providers.

Reply:
I would think returning not supported would be sufficient..

Reply:
It's likely the new API is the new normal. All they NIST signature algorithms are defined needing the signature first. It provides better signature guarantees. Protocol people don't like it because protocols often put the signature at the end. My guess is protocols with end up doing single shot where they use the full algorithm to sign their preselected hashes.

Reply:
I disagree